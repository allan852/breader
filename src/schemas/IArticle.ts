export default interface IArtilce {
    author: string
    // categories?: string[]
    // comments?: string
    created_at?: number
    date: string
    feed_id?: number
    time: string
    description?: string
    feed_title?: string
    // enclosures?: InterfaceFeedEnclosure[]
    guid: string
    // image?: InterfaceFeedImage
    id?: number
    index?: number
    is_starred?: number
    is_unread?: number
    is_dayfirst?: boolean
    link: string
    // origlink?: string
    // permalink?: string
    // pubdate: string
    // source?: InterfaceFeedSource
    summary: string
    title: string
    updated_at?: number
};
