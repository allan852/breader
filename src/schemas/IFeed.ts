export default interface IFeed {
    id?: number
    title?: string
    link: string
    url: string
    date?: Date
    date_time?: string
    etag?: string
    description?: string
    favicon?: string
    summary?: string
    deleted_at?: number
}
