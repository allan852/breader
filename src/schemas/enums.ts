export enum ELogicError {
  FEEDPARSER_FETCH_ERROR = 'FEEDPARSER_FETCH_ERROR',
  FEEDPARSER_NOT_FOUND = 'FEEDPARSER_NOT_FOUND',
  FEEDPARSER_WRONG_URL = 'FEEDPARSER_WRONG_URL',
  POUCHDB_EXISTS = 'POUCHDB_EXISTS',
  UNKNOWN = 'UNKNOWN',
}

export enum EMenuKey {
  ALL_ITEMS = 'ALL_ITEMS',
  STARRED_ITEMS = 'STARRED_ITEMS',
  UNREAD_ITEMS = 'UNREAD_ITEMS',
}

export enum EArticleFilter {
  STARRED = 'STARRED',
  UNREAD = 'UNREAD',
  ALL = 'ALL',
}
