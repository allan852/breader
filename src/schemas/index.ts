export * from './enums'
export * from './IArticle'
export * from './IFeed'
export * from './IReduxAction'
export * from './IReduxAsyncAction'
export * from './IReduxState'
