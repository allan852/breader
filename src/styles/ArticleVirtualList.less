@import './Values.less';

.article-virtual-list {
  height: 100%;
  position: relative;

  &:focus,
  * {
    outline: none;
    outline-color: @beige-background-color;
  }
  .date-divid {
    color: @black-text-color-dark;
    border-bottom: none;
    font-size: 12px;
    padding: 0 8px 0 8px;
    margin-top: 24px;
  }
  .first-list-item .date-divid {
    margin-top: 0;
  }

  .list-affix {
    width: 100%;
    position: absolute;
    top: 0;
    font-size: 12px;
    color: @black-text-color-dark;
    z-index: 999;
    padding: 2px 8px 2px 8px;
    background-color: @grey-dividing-color;
    border-top: 0.5px solid @grey-border-color;
    border-bottom: 0.5px solid @grey-border-color;
  }
}
