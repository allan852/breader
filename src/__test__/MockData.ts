import { IntlProvider } from 'react-intl'
import { messages } from '../locales'
import { IArticle, IFeed } from '../schemas'

export const intlProviderProps: IntlProvider.Props = {
  locale: 'en-US',
  messages: messages['en-US'],
}

const intlProvider = new IntlProvider(intlProviderProps, {})
export const intl = intlProvider.getChildContext().intl

export const feed: IFeed = {
  _id: 'http://www.ruanyifeng.com/blog/atom.xml',
  _rev: '5-55772bf2b4d4b74608a60ac3480334e4',
  author: 'ruanyifeng',
  categories: [],
  createTime: 1563686468437,
  deleteTime: 0,
  description: "Ruan YiFeng's Blog",
  etag: '"11d86-58e2cc1eab031"',
  favicon: 'http://www.ruanyifeng.com/favicon.ico',
  generator: 'Movable Type 5.2.2 (http://www.sixapart.com/movabletype/)',
  language: '',
  link: 'http://www.ruanyifeng.com/blog/',
  publishTime: 1563697765000,
  time: 1563697765000,
  title: '阮一峰的网络日志',
  url: 'http://www.ruanyifeng.com/blog/atom.xml',
}

export const article: IArticle = {
  _id: 'tag:www.ruanyifeng.com,2019:/blog//1.2100',
  _rev: '5-ff6fc450616071a8f7d022b7532e85aa',
  author: 'ruanyifeng',
  categories: [],
  comments: 'article comments',
  createTime: 1563697765000,
  deleteTime: 0,
  description: 'article description',
  enclosures: [],
  feedId: 'http://www.ruanyifeng.com/blog/atom.xml',
  image: '',
  isDayFirst: true,
  isStarred: false,
  isUnread: false,
  link: 'http://www.ruanyifeng.com/blog/2019/07/weekly-issue-65.html',
  originLink: '',
  publishTime: 1563697765000,
  summary: 'article summary',
  time: 1563697765000,
  title: 'article title',
}
