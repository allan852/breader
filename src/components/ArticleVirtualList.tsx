import { Empty } from 'antd'
import Immutable from 'immutable'
import React, { PureComponent, RefObject } from 'react'
import {
  AutoSizer,
  CellMeasurer,
  CellMeasurerCache,
  List as VList,
} from 'react-virtualized'
import ArticleItem from '../containers/ArticleItem'
import { IArticle } from '../schemas'
import '../styles/ArticleVirtualList.less'
import Utils from '../utils'

export interface IArticleVirtualListOwnProps {
  scrollToIndex?: number
}

export interface IArticleVirtualListReduxDispatch {
  selectArticle: (id: string, index: number) => Promise<IArticle | null>
}

export interface IArticleVirtualListReduxState {
  currentArticle: IArticle | null
  articles: Immutable.List<IArticle>
}

interface IArticleVirtualListProps
  extends IArticleVirtualListOwnProps,
    IArticleVirtualListReduxDispatch,
    IArticleVirtualListReduxState {}

interface IArticleVirtualListState {
  cellCache: CellMeasurerCache
  isAffixVisible: boolean
  lastArticles: Immutable.List<IArticle>
  readItems: { [_id: string]: boolean }
  renderStartDate: string
}

class ArticleVirtualList extends PureComponent<
  IArticleVirtualListProps,
  IArticleVirtualListState
> {
  public static getDerivedStateFromProps(
    nextProps: IArticleVirtualListProps,
    prevState: IArticleVirtualListState
  ) {
    if (nextProps.articles !== prevState.lastArticles) {
      return {
        cellCache: new CellMeasurerCache({
          defaultHeight: 80,
          fixedWidth: true,
        }),
        lastArticles: nextProps.articles,
      }
    }
    return null
  }
  public vlist: RefObject<VList>
  public updateRenderStartDate: (...params: any[]) => void
  public constructor(props: IArticleVirtualListProps) {
    super(props)
    this.vlist = React.createRef()
    this.state = {
      cellCache: new CellMeasurerCache({
        defaultHeight: 80,
        fixedWidth: true,
      }),
      isAffixVisible: false,
      lastArticles: props.articles,
      readItems: {},
      renderStartDate: '',
    }
    this.updateRenderStartDate = Utils.throttle(
      this._updateRenderStartDate,
      100
    )

    // Object.defineProperty(window, 'updateRenderStartDate', {value: this.updateRenderStartDate})
  }
  public componentDidUpdate(prevProps: IArticleVirtualListProps) {
    const props = this.props
    if (
      props.scrollToIndex !== undefined &&
      props.scrollToIndex > -1 &&
      props.scrollToIndex !== prevProps.scrollToIndex
    ) {
      const vlist = this.vlist.current
      const index = props.scrollToIndex
      if (vlist) {
        vlist.scrollToRow(props.scrollToIndex)
        setImmediate(() => {
          const items = document.querySelectorAll('.vlist-item')
          items.forEach((item: any) => {
            if (
              item.dataset.index === index ||
              item.dataset.index === index + ''
            ) {
              item.click()
            }
          })
        })
      }
    }
  }
  public handleClick = (e: React.MouseEvent<HTMLDivElement>) => {
    const readItems = this.state.readItems
    const target = e.target
    const $item = (target as HTMLDivElement).closest('.vlist-item')
    if ($item) {
      const { id, index } = ($item as HTMLDivElement).dataset
      const article = this.props.currentArticle
      if (article && article._id === id) {
        // do nothing
      } else if (id && index) {
        this.props.selectArticle(id, parseInt(index, 10))
        readItems[id] = true
        this.setState({
          readItems: { ...readItems },
        })
      }
    }
  }
  public render() {
    return (
      <div className="article-virtual-list" onClick={this.handleClick}>
        {this.state.isAffixVisible && (
          <div className="list-affix">{this.state.renderStartDate}</div>
        )}
        <AutoSizer>
          {({ width, height }) => (
            <VList
              ref={this.vlist}
              width={width}
              height={height}
              deferredMeasurementCache={this.state.cellCache}
              rowCount={this.props.articles.size}
              rowHeight={this.state.cellCache.rowHeight}
              rowRenderer={(info: any) => this._rowRenderer(info)}
              noRowsRenderer={() => this._noRowsRenderer()}
              onScroll={(info: any) => this._onScroll(info)}
            />
          )}
        </AutoSizer>
      </div>
    )
  }
  private _onScroll(info: any) {
    if (info.scrollTop > 3 && this.state.isAffixVisible === false) {
      this.setState({
        isAffixVisible: true,
      })
    }
    if (info.scrollTop <= 3 && this.state.isAffixVisible === true) {
      this.setState({
        isAffixVisible: false,
      })
    }
    this.updateRenderStartDate.call(this)
  }
  private _updateRenderStartDate() {
    const vlist = this.vlist.current
    if (vlist && vlist.Grid) {
      const startIndex = (vlist.Grid as any)._renderedRowStartIndex
      const startArticle = this.props.articles.get(startIndex)
      if (startArticle) {
        const dateTime = Utils.timeToDateString(startArticle.time)
        if (dateTime !== this.state.renderStartDate) {
          this.setState({
            renderStartDate: dateTime,
          })
        }
      }
    }
  }
  private _noRowsRenderer() {
    return (
      <div style={{ marginTop: '30px' }}>
        <Empty />
      </div>
    )
  }
  private _rowRenderer(info: any) {
    const index = info.index
    const parent = info.parent
    const key = info.key
    const style = info.style
    const article = this.props.articles.get(index) as IArticle
    const currentArticle = this.props.currentArticle
    const isCurrent = currentArticle && article._id === currentArticle._id
    return (
      <CellMeasurer
        key={key}
        cache={this.state.cellCache}
        parent={parent}
        columnIndex={0}
        rowIndex={index}>
        <div
          style={style}
          data-id={article._id}
          data-index={index}
          className={index === 0 ? 'vlist-item first-list-item' : 'vlist-item'}>
          {article.isDayFirst && (
            <div className="date-divid">
              {Utils.timeToDateString(article.time)}
            </div>
          )}
          <ArticleItem
            author={article.author}
            guid={article._id}
            time={article.time}
            feedId={article.feedId}
            title={article.title}
            summary={article.summary}
            className={
              (article.isUnread && !this.state.readItems[article._id]
                ? 'item-is-unread'
                : '') + (isCurrent ? ' item-is-selected' : '')
            }
          />
        </div>
      </CellMeasurer>
    )
  }
}

export default ArticleVirtualList
