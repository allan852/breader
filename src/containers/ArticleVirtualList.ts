import { connect, MapDispatchToProps, MapStateToProps } from 'react-redux'
import { Dispatch } from 'redux'
import ArticleVirtualList, {
  IArticleVirtualListDispatchProps,
  IArticleVirtualListOwnProps,
  IArticleVirtualListStateProps,
} from '../components/ArticleVirtualList'
import {
  asyncActionDispatcher,
  asyncSelectAndReadArticleAction,
} from '../redux/actions'
import { IReduxAction, IReduxState } from '../schemas'

const mapStateToProps: MapStateToProps<
  IArticleVirtualListStateProps,
  IArticleVirtualListOwnProps,
  IReduxState
> = (state: IReduxState) => ({
  articles: state.articles.list,
  currentArticle: state.articles.current,
  selectedMenuKey: state.menu.selectedKey,
})

const mapDispatchToProps: MapDispatchToProps<
  IArticleVirtualListDispatchProps,
  IArticleVirtualListOwnProps
> = (dispatch: Dispatch<IReduxAction>) => ({
  selectArticle: (id: string, index: number) =>
    asyncActionDispatcher(dispatch, asyncSelectAndReadArticleAction(id, index)),
})

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(ArticleVirtualList)
