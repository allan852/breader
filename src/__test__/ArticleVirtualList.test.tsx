import Enzyme from 'enzyme'
import EnzymeAdapter from 'enzyme-adapter-react-16'
import EnzymeToJson from 'enzyme-to-json'
import Immutable from 'immutable'
import React from 'react'
import { IntlProvider } from 'react-intl'
import ArticleVirtualList, {
  IArticleVirtualListProps,
} from '../components/ArticleVirtualList'
import { EMenuKey, IArticle } from '../schemas'
import { article, intlProviderProps } from './MockData'

Enzyme.configure({ adapter: new EnzymeAdapter() })

describe('ArticleVirtualList Testing', () => {
  const mockProps: IArticleVirtualListProps = {
    articles: Immutable.List<IArticle>([article]),
    currentArticle: null,
    scrollToIndex: 0,
    selectArticle: jest.fn(),
    selectedMenuKey: EMenuKey.ALL_ITEMS,
  }

  it('dom testing', () => {
    const wrapper = Enzyme.mount(<ArticleVirtualList {...mockProps} />, {
      wrappingComponent: IntlProvider,
      wrappingComponentProps: intlProviderProps,
    })
    wrapper.setProps({
      currentArticle: article,
    })
    wrapper.update()
  })

  it('snapshot testing', () => {
    const wrapper = Enzyme.mount(<ArticleVirtualList {...mockProps} />, {
      wrappingComponent: IntlProvider,
      wrappingComponentProps: intlProviderProps,
    })
    const tree = EnzymeToJson(wrapper)
    expect(tree).toMatchSnapshot()
  })
})
