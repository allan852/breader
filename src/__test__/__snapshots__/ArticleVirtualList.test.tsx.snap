// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArticleVirtualList Testing snapshot testing 1`] = `
<ArticleVirtualList
  articles={
    Immutable.List [
      Object {
        "_id": "tag:www.ruanyifeng.com,2019:/blog//1.2100",
        "_rev": "5-ff6fc450616071a8f7d022b7532e85aa",
        "author": "ruanyifeng",
        "categories": Array [],
        "comments": "article comments",
        "createTime": 1563697765000,
        "deleteTime": 0,
        "description": "article description",
        "enclosures": Array [],
        "feedId": "http://www.ruanyifeng.com/blog/atom.xml",
        "image": "",
        "isDayFirst": true,
        "isStarred": false,
        "isUnread": false,
        "link": "http://www.ruanyifeng.com/blog/2019/07/weekly-issue-65.html",
        "originLink": "",
        "publishTime": 1563697765000,
        "summary": "article summary",
        "time": 1563697765000,
        "title": "article title",
      },
    ]
  }
  currentArticle={null}
  scrollToIndex={0}
  selectArticle={[MockFunction]}
  selectedMenuKey="ALL_ITEMS"
>
  <div
    className="article-virtual-list"
    onClick={[Function]}
  >
    <Scrollbars
      autoHeight={false}
      autoHeightMax={200}
      autoHeightMin={0}
      autoHide={false}
      autoHideDuration={200}
      autoHideTimeout={1000}
      hideTracksWhenNotNeeded={false}
      renderThumbHorizontal={[Function]}
      renderThumbVertical={[Function]}
      renderTrackHorizontal={[Function]}
      renderTrackVertical={[Function]}
      renderView={[Function]}
      tagName="div"
      thumbMinSize={30}
      universal={false}
    >
      <div
        style={
          Object {
            "height": "100%",
            "overflow": "hidden",
            "position": "relative",
            "width": "100%",
          }
        }
      >
        <div
          key="view"
          style={
            Object {
              "WebkitOverflowScrolling": "touch",
              "bottom": 0,
              "left": 0,
              "marginBottom": 0,
              "marginRight": 0,
              "overflow": "scroll",
              "position": "absolute",
              "right": 0,
              "top": 0,
            }
          }
        >
          <AutoSizer
            disableHeight={false}
            disableWidth={false}
            onResize={[Function]}
            style={Object {}}
          >
            <div
              style={
                Object {
                  "height": 0,
                  "overflow": "visible",
                  "width": 0,
                }
              }
            >
              <List
                autoHeight={false}
                deferredMeasurementCache={
                  CellMeasurerCache {
                    "_cellHeightCache": Object {},
                    "_cellWidthCache": Object {},
                    "_columnCount": 0,
                    "_columnWidthCache": Object {},
                    "_defaultHeight": 80,
                    "_defaultWidth": 100,
                    "_hasFixedHeight": false,
                    "_hasFixedWidth": true,
                    "_keyMapper": [Function],
                    "_minHeight": 0,
                    "_minWidth": 0,
                    "_rowCount": 0,
                    "_rowHeightCache": Object {},
                    "columnWidth": [Function],
                    "rowHeight": [Function],
                  }
                }
                estimatedRowSize={30}
                height={0}
                noRowsRenderer={[Function]}
                onRowsRendered={[Function]}
                onScroll={[Function]}
                overscanIndicesGetter={[Function]}
                overscanRowCount={10}
                rowCount={1}
                rowHeight={[Function]}
                rowRenderer={[Function]}
                scrollToAlignment="auto"
                scrollToIndex={-1}
                style={Object {}}
                width={0}
              >
                <Grid
                  aria-label="grid"
                  aria-readonly={true}
                  autoContainerWidth={true}
                  autoHeight={false}
                  autoWidth={false}
                  cellRangeRenderer={[Function]}
                  cellRenderer={[Function]}
                  className="ReactVirtualized__List"
                  columnCount={1}
                  columnWidth={0}
                  containerRole="rowgroup"
                  containerStyle={Object {}}
                  deferredMeasurementCache={
                    CellMeasurerCache {
                      "_cellHeightCache": Object {},
                      "_cellWidthCache": Object {},
                      "_columnCount": 0,
                      "_columnWidthCache": Object {},
                      "_defaultHeight": 80,
                      "_defaultWidth": 100,
                      "_hasFixedHeight": false,
                      "_hasFixedWidth": true,
                      "_keyMapper": [Function],
                      "_minHeight": 0,
                      "_minWidth": 0,
                      "_rowCount": 0,
                      "_rowHeightCache": Object {},
                      "columnWidth": [Function],
                      "rowHeight": [Function],
                    }
                  }
                  estimatedColumnSize={100}
                  estimatedRowSize={30}
                  getScrollbarSize={[Function]}
                  height={0}
                  isScrollingOptOut={false}
                  noContentRenderer={[Function]}
                  noRowsRenderer={[Function]}
                  onRowsRendered={[Function]}
                  onScroll={[Function]}
                  onScrollbarPresenceChange={[Function]}
                  onSectionRendered={[Function]}
                  overscanColumnCount={0}
                  overscanIndicesGetter={[Function]}
                  overscanRowCount={10}
                  role="grid"
                  rowCount={1}
                  rowHeight={[Function]}
                  rowRenderer={[Function]}
                  scrollToAlignment="auto"
                  scrollToColumn={-1}
                  scrollToIndex={-1}
                  scrollToRow={-1}
                  scrollingResetTimeInterval={150}
                  style={Object {}}
                  tabIndex={0}
                  width={0}
                >
                  <div
                    aria-label="grid"
                    aria-readonly={true}
                    className="ReactVirtualized__Grid ReactVirtualized__List"
                    onScroll={[Function]}
                    role="grid"
                    style={
                      Object {
                        "WebkitOverflowScrolling": "touch",
                        "boxSizing": "border-box",
                        "direction": "ltr",
                        "height": 0,
                        "overflowX": "hidden",
                        "overflowY": "auto",
                        "position": "relative",
                        "width": 0,
                        "willChange": "transform",
                      }
                    }
                    tabIndex={0}
                  />
                </Grid>
              </List>
            </div>
          </AutoSizer>
        </div>
        <div
          key="trackHorizontal"
          style={
            Object {
              "borderRadius": 3,
              "bottom": 2,
              "display": "none",
              "height": 6,
              "left": 2,
              "position": "absolute",
              "right": 2,
            }
          }
        >
          <div
            style={
              Object {
                "backgroundColor": "rgba(0,0,0,.2)",
                "borderRadius": "inherit",
                "cursor": "pointer",
                "display": "block",
                "height": "100%",
                "position": "relative",
              }
            }
          />
        </div>
        <div
          key="trackVertical"
          style={
            Object {
              "borderRadius": 3,
              "bottom": 2,
              "display": "none",
              "position": "absolute",
              "right": 2,
              "top": 2,
              "width": 6,
            }
          }
        >
          <div
            style={
              Object {
                "backgroundColor": "rgba(0,0,0,.2)",
                "borderRadius": "inherit",
                "cursor": "pointer",
                "display": "block",
                "position": "relative",
                "width": "100%",
              }
            }
          />
        </div>
      </div>
    </Scrollbars>
  </div>
</ArticleVirtualList>
`;
